canary/archival.go

```diff
@@ -158,6 +190,85 @@ func archivalExternalWorkflow(ctx workflow.Context, scheduledTimeNanos int64) er
 	return profile.end(nil)
 }
 
-func archivalLargeResultActivity() ([]byte, error) {
+func largeResultActivity() ([]byte, error) {
 	return make([]byte, resultSize, resultSize), nil
 }
+
+func visibilityArchivalActivity(ctx context.Context, scheduledTimeNanos int64) error {
+	scope := activity.GetMetricsScope(ctx)
+	var err error
+	scope, sw := recordActivityStart(scope, activityTypeVisibilityArchival, scheduledTimeNanos)
+	defer recordActivityEnd(scope, sw, err)
+
+	client := getActivityArchivalContext(ctx).cadence
+	resp, err := client.Describe(ctx, archivalDomain)
+	if err != nil {
+		return err
+	}
+
+	if resp.Configuration != nil &&
+		resp.Configuration.GetVisibilityArchivalStatus() == shared.ArchivalStatusDisabled {
+		return errors.New("domain not configured for visibility archival")
+	}
+
+	visArchivalURI := ""
+	if resp.Configuration != nil {
+		visArchivalURI = resp.Configuration.GetVisibilityArchivalURI()
+	}
+
+	scheme := getURIScheme(visArchivalURI)
+	if scheme != visArchivalScheme {
+		return fmt.Errorf("unknown visibility archival scheme: %s, expecting %s", scheme, visArchivalScheme)
+	}
+
+	listReq := &shared.ListArchivedWorkflowExecutionsRequest{
+		Domain:   stringPtr(archivalDomain),
+		PageSize: int32Ptr(visArchivalPageSize),
+		Query: stringPtr(fmt.Sprintf(visArchivalQueryTemplate,
+			wfTypeArchival,
+			time.Now().Add(-time.Hour*12).UnixNano(),
+			time.Now().Add(-time.Hour*6).UnixNano(),
+		)),
+	}
+	listResp := &shared.ListArchivedWorkflowExecutionsResponse{}
+	var executions []*shared.WorkflowExecutionInfo
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		scope.Counter(listArchivedWorkflowCount).Inc(1)
+		sw := scope.Timer(listArchivedWorkflowsLatency).Start()
+		listResp, err = client.ListArchivedWorkflow(ctx, listReq)
+		sw.Stop()
+		if err != nil {
+			scope.Counter(listArchivedWorkflowFailureCount).Inc(1)
+			return err
+		}
+
+		if len(listResp.Executions) != 0 {
+			executions = append(executions, listResp.Executions...)
+		}
+
+		if listResp.NextPageToken == nil {
+			break
+		}
+
+		listReq.NextPageToken = listResp.NextPageToken
+	}
+
+	if len(listResp.Executions) == 0 {
```

Comment:
Roughly how many do we expect to have here? Seems like if we can assert something stronger than more than 0 it would be good.

Reply:
I put more thought on this last night and made the following two changes:
1. Separate the canary test for history archival and visibility archival so they can enabled and disabled independently.
2. Created an interface for validating visibility archival (a template) which each visibility archival scheme should implement. It basically tells canary if the test should be run, what is the query and how to validate the result. With this interface I believe the code is general enough for all archiver implementations and each implementation can have more control over the test. 

I made those changes because I think when open source customer uses canary in production, it's unlikely they set the scheme for canary domain to filestore, it will be either s3 or gcp or other implementation. So we should make the code more general and flexible. 

What's your thought on this?

Reply:
This makes a lot of sense :)